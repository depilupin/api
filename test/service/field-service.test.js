const Contracts = require('../../src/datasources/contracts/contracts');
const FieldService = require('../../src/service/field-service');
const { toBigInt } = require('../../src/utils/number');
const { mockBeanstalkSG } = require('../util/mock-sg');

describe('FieldService', () => {
  describe('Plot summary', () => {
    beforeEach(async () => {
      jest.restoreAllMocks();
      const allPlots = require('../mock-responses/service/field/allPlots.json');
      jest.spyOn(mockBeanstalkSG, 'request').mockResolvedValueOnce(allPlots);
      jest.spyOn(Contracts, 'getBeanstalk').mockReturnValue({
        harvestableIndex: jest.fn().mockResolvedValue(5000)
      });
    });

    it('Small plots overlapping 2 buckets', async () => {
      const result = await FieldService.getAggregatePlotSummary({ bucketSize: 10000 });

      expect(result.length).toBe(3);
      expect(result[0].startIndex).toBe(0n);
      expect(result[0].endIndex).toBe(toBigInt(10000, 6));
      expect(result[0].startSeason).toBe(2);
      expect(result[0].endSeason).toBe(5);
      expect(result[0].numPlots).toBe(2);
      expect(result[2].endIndex).toBe(toBigInt(23500, 6));

      expect(result[0].avgSownBeansPerPod).toBeCloseTo(0.525, 3);
      expect(result[1].avgSownBeansPerPod).toBeCloseTo(0.3, 3);
      expect(result[2].avgSownBeansPerPod).toBeCloseTo(0.2, 3);

      expect(result.map((x) => x.avgAPR)).toMatchSnapshot();
    });

    it('Small plots starting/ending at bucket boundaries', async () => {
      const result = await FieldService.getAggregatePlotSummary({ bucketSize: 7500 });

      expect(result.length).toBe(4);
      expect(result[0].startIndex).toBe(0n);
      expect(result[0].endIndex).toBe(toBigInt(7500, 6));
      expect(result[0].startSeason).toBe(2);
      expect(result[0].endSeason).toBe(2);
      expect(result[0].numPlots).toBe(1);
      expect(result[1].startSeason).toBe(5);
      expect(result[1].numPlots).toBe(2);
      expect(result[3].endIndex).toBe(toBigInt(23500, 6));

      expect(result[0].avgSownBeansPerPod).toBeCloseTo(0.5, 3);
      expect(result[1].avgSownBeansPerPod).toBeCloseTo(0.467, 3);
      expect(result[2].avgSownBeansPerPod).toBeCloseTo(0.2, 3);
      expect(result[3].avgSownBeansPerPod).toBeCloseTo(0.2, 3);

      expect(result.map((x) => x.avgAPR)).toMatchSnapshot();
    });

    it('Large plots filling entire buckets', async () => {
      const result = await FieldService.getAggregatePlotSummary({ bucketSize: 3000 });

      expect(result.length).toBe(8);
      expect(result[0].numPlots).toBe(1);
      expect(result[1].numPlots).toBe(1);
      expect(result[2].numPlots).toBe(2);

      expect(result[0].avgSownBeansPerPod).toBeCloseTo(0.5, 3);
      expect(result[1].avgSownBeansPerPod).toBeCloseTo(0.5, 3);
      expect(result[2].avgSownBeansPerPod).toBeCloseTo(0.55, 3);

      expect(result.map((x) => x.avgAPR)).toMatchSnapshot();
    });

    describe('Cache/time manipulation', () => {
      beforeAll(() => {
        jest.useFakeTimers();
      });

      afterAll(() => {
        jest.useRealTimers();
      });

      it('Uses cached values when available', async () => {
        const timestamp = new Date();
        jest.spyOn(FieldService, 'cache', 'get').mockReturnValue({
          10000: {
            timestamp,
            result: 'result'
          }
        });
        const sgSpy = jest.spyOn(mockBeanstalkSG, 'request');

        const result = await FieldService.getAggregatePlotSummary({ bucketSize: 10000 });

        expect(result).toBe('result');
        expect(sgSpy).not.toHaveBeenCalled();
      });

      it('Cache miss due to stale timestamp', async () => {
        jest.spyOn(FieldService, 'cache', 'get').mockReturnValue({
          10000: {
            timestamp: new Date(Date.now() - 1000 * 60 * 300 - 1),
            result: 'result'
          }
        });
        const sgSpy = jest.spyOn(mockBeanstalkSG, 'request');

        jest.setSystemTime(Date.now() + 1000 * 60 * 300 + 1);
        const result = await FieldService.getAggregatePlotSummary({ bucketSize: 10000 });

        expect(result).not.toBe('result');
        expect(sgSpy).toHaveBeenCalled();
      });

      it('Does not cache values for random bucket sizes', async () => {
        await FieldService.getAggregatePlotSummary({ bucketSize: 12345 });

        expect(FieldService.cache[12345]).toBeUndefined();
      });
    });
  });
});
