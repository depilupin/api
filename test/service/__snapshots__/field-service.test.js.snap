// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldService Plot summary Large plots filling entire buckets 1`] = `
[
  10.512,
  10.512,
  5.256,
  undefined,
  undefined,
  undefined,
  undefined,
  undefined,
]
`;

exports[`FieldService Plot summary Small plots overlapping 2 buckets 1`] = `
[
  7.884,
  undefined,
  undefined,
]
`;

exports[`FieldService Plot summary Small plots starting/ending at bucket boundaries 1`] = `
[
  10.512,
  undefined,
  undefined,
  undefined,
]
`;
